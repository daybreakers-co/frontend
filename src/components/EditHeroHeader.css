.EditHeroHeader input.title {
  font-size: 4.6rem;
  line-height: 1.2;
  color: white;
  text-align: center;
  background: transparent;
  text-shadow: 0 1px 0px rgba(0,0,0,0.4);
  appearance: none;
  box-shadow: none;
  border-radius: none;
  outline: none;
}

.EditHeroHeader input.subtitle {
  color: white;
  text-align: center;
  background: transparent;
  text-shadow: 0 1px 0px rgba(0,0,0,0.4);
  appearance: none;
  box-shadow: none;
  border-radius: none;
  outline: none;
}

.EditHeroHeader {
  width: 100%;
  height: 80vh;
  background-color: #bdbdbd;
  color: white;
  text-shadow: 0 1px 0px rgba(0,0,0,0.4);
  position: relative;
}

.EditHeroHeader figure {
  height: 80vh;
}

.EditHeroHeader img {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.EditHeroHeader hgroup {
  background-color: rgba(0,0,0,0.4);
  height: 100%;
  width: 100%;
  padding-top: 30vh;
  z-index: 100;
  top: 0;
  position: absolute;
}
