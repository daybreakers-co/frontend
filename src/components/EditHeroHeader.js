import React from 'react'
import PropTypes from 'prop-types'
import Dropzone from 'react-dropzone'
import UploadableImage from './UploadableImage'
import ScaledImage from './ScaledImage'

import './HeroHeader.css'
import './EditHeroHeader.css'

class EditHeroHeader extends React.Component {
  constructor(props) {
    super()
    this.state = {
      title: props.title,
      subtitle: props.subtitle,
      file: null
    }
  }

  static propTypes = {
    title: PropTypes.string,
    subtitle: PropTypes.string,
    header: PropTypes.object,
    type: PropTypes.string,
    uploadParentId: PropTypes.string.isRequired,
    uploadParentType: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      title: nextProps.title,
      subtitle: nextProps.subtitle,
      header: nextProps.header
    })
  }

  onDrop = (files) => {
    this.setState({file: files[0]})
  }

  handleBlur = () => {
    this.props.onChange({
      title:    this.state.title,
      subtitle: this.state.subtitle
    })
  }

  render () {
    const { header, uploadParentId, uploadParentType, type } = this.props
    const { title, subtitle, file } = this.state
    let backgroundImage;

    if (file) {
      backgroundImage = <UploadableImage
      parentType={uploadParentType}
      parentId={uploadParentId}
      file={this.state.file} />
    } else if (header) {
      backgroundImage = <ScaledImage
      key={header.id}
      image={header}
      style={{flex: header.ratio}}
      alt="image"/>
    }
    return (
      <Dropzone onDrop={this.onDrop} className={`HeroHeader edit ${type}`} disableClick={true}>
        {backgroundImage}
        <hgroup>
          <input
            className="title"
            value={title || ""}
            placeholder='(Amazing title)'
            onChange={(e) => this.setState({title: e.target.value})}
            onBlur={this.handleBlur} />
          <input
            className="subtitle"
            value={subtitle || ""}
            placeholder='(Optional subtitle)'
            onChange={(e) => this.setState({subtitle: e.target.value})}
            onBlur={this.handleBlur} />

          <p className="drophint"><i className="fa fa-picture-o" /> Drop an image here to update the header image.</p>
        </hgroup>
      </Dropzone>
    )
  }
}

export default EditHeroHeader;
