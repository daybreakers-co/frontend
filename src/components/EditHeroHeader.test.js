import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json';

import EditHeroHeader from './EditHeroHeader'

describe("EditHeroHeader", () => {
  const photo = {id: "def234", ratio: 1, url: "image.jpg"}
  const file = {}
  const spy = jest.fn()

  describe("With photo", () => {
    it("Renders a scaled image", () => {
      const wrapper = shallow(
        <EditHeroHeader
          title="Header title"
          subtitle="Header subtitle"
          uploadParentId="trip1"
          uploadParentType="Trip"
          header={photo}
          onChange={spy} />
      )

      let scaledImage = wrapper.find("ScaledImage")
      expect(scaledImage.prop("image")).toEqual(photo)
      expect(scaledImage.prop("style")).toEqual({flex: 1})
    })
  })

  describe("With file", () => {
    it("Renders an UploadableImage", () => {
      const wrapper = shallow(
        <EditHeroHeader
          title="Header title"
          subtitle="Header subtitle"
          uploadParentId="trip1"
          uploadParentType="Trip"
          header={photo}
          onChange={spy} />
      )
      wrapper.setState({file: file})

      let uploadableImage = wrapper.find("UploadableImage")
      expect(uploadableImage.prop("parentId")).toEqual("trip1")
      expect(uploadableImage.prop("parentType")).toEqual("Trip")
      expect(uploadableImage.prop("file")).toEqual(file)
    })
  })

  it("Uploads an image when file is dropped", () => {
    const wrapper = shallow(
      <EditHeroHeader
        title="Header title"
        subtitle="Header subtitle"
        uploadParentId="trip1"
        uploadParentType="Trip"
        header={photo}
        onChange={spy} />
    )
    wrapper.instance().onDrop.call(wrapper.instance(), [file])

    expect(wrapper.instance().state.file).toEqual(file)
  })

  it("Calls onChange when input is blurred", () => {
    const wrapper = shallow(
      <EditHeroHeader
        title="Header title"
        subtitle="Header subtitle"
        uploadParentId="trip1"
        uploadParentType="Trip"
        header={photo}
        onChange={spy} />
    )

    // Title input
    wrapper.find("input").at(0).simulate("blur", {})
    expect(spy).toBeCalledWith({title: "Header title", subtitle: "Header subtitle"})

    wrapper.setState({title: "New title", subtitle: "New subtitle"})

    // Subtitle input
    wrapper.find("input").at(1).simulate("blur", {})
    expect(spy).toBeCalledWith({title: "New title", subtitle: "New subtitle"})
  })

  it("Updates state from new props", () => {
    const wrapper = shallow(
      <EditHeroHeader
        title="Header title"
        subtitle="Header subtitle"
        uploadParentId="trip1"
        uploadParentType="Trip"
        header={photo}
        onChange={spy} />
    )

    wrapper.instance().componentWillReceiveProps.call(wrapper.instance(), {
      title: "New title",
      subtitle: "New subtitle",
      header: photo
    })

    expect(wrapper.instance().state).toEqual({
      title: "New title",
      subtitle: "New subtitle",
      header: photo,
      file: null
    })
  })

  it ("Matches snapshot",() => {
    const wrapper = shallow(
      <EditHeroHeader
        title="Header title"
        subtitle="Header subtitle"
        uploadParentId="trip1"
        uploadParentType="Trip"
        header={photo}
        onChange={spy} />
    )

    expect(toJson(wrapper)).toMatchSnapshot();
  })
})
