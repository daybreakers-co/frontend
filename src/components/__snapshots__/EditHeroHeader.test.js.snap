// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditHeroHeader Matches snapshot 1`] = `
<Dropzone
  className="HeroHeader edit undefined"
  disableClick={true}
  disablePreview={false}
  maxSize={Infinity}
  minSize={0}
  multiple={true}
  onDrop={[Function]}
  preventDropOnDocument={true}
>
  <ScaledImage
    alt="image"
    cover={true}
    image={
      Object {
        "id": "def234",
        "ratio": 1,
        "url": "image.jpg",
      }
    }
    key="def234"
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <hgroup>
    <input
      className="title"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="(Amazing title)"
      value="Header title"
    />
    <input
      className="subtitle"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="(Optional subtitle)"
      value="Header subtitle"
    />
    <p
      className="drophint"
    >
      <i
        className="fa fa-picture-o"
      />
       Drop an image here to update the header image.
    </p>
  </hgroup>
</Dropzone>
`;
